Automattic/wp-calypso
# WIP: Add behaviour for dealing with duplicate sites in Gutenboarding

***Note:*** This PR originally solved the issue by adding a modal. The PR has since been simplified to deal with redirect behaviour implicitly.

---

This PR is an attempt at solving the duplicate sites in signup issue for Gutenboarding by storing the domain of the last created site in state. When Gutenboarding loads, if the last created site is recent (within 60 minutes) and the user is not on the first step of the flow (e.g. they landed there by hitting the back button), then we redirect to the customer home. If instead they navigated to the beginning of the flow (e.g. they clicked "Add a new site"), we clear the state and begin a fresh flow.

After the 60 minutes have elapsed, if a user navigates to Gutenboarding and a last created site exists in state, the Gutenboarding state is reset, and the user is redirected to the beginning of the signup flow, irrespective of which step they were on.

#### Changes proposed in this Pull Request

* Add logic to the Gutenboard component, that redirects the user to the Customer Home given the following conditions:
  * Haven't checked for a potential duplicate site yet
  * The user is logged in (there is a `currentUser`)
  * The last site was created recently (within the past 60 minutes)
  * The site exists
  * The user landed on a part of Gutenboarding that is _not_ the first step (e.g. they reached Gutenboarding via hitting the Back button)

* If the user begins at the first step of Gutenboarding (e.g. from clicking "Add a site") or the last site wasn't created recently, then we clear the onboarding store (and last created site state) and start from the beginning.

* Add a selector and resolver to the `site` data store `getSite` to pass in a domain/slug string and check against the `/sites` endpoint whether or not the site exists. This stores sites in an object, keyed by the site's slug.

#### Animated gif

Going through the following steps:

* From a freshly created site, the user clicks back
* They land on the last step of the Gutenboarding flow, and are then redirected to the Customer Home

![duplicate-site-redirect](https://user-images.githubusercontent.com/14988353/76391027-a74f0380-63c2-11ea-8555-f32e52059f7e.gif)

#### Testing instructions



* Go to `/gutenboarding` and create a new site.
* Click back and you should be taken to the Customer Home for that site.
* Also create a new site, and then manually navigate to `/gutenboarding`, simulating what it might look like for a user who creates a site and then goes to "Add a new site" from the Calypso admin — it should begin a new site creation flow like normal.

Fixes #39268
